README
This should be a temporary fork of Math-Clipper with early support for points
with Z values. This is a new compile-time option introduced in 
Clipper 6.0.0 alpha.

Ideally Math::Clipper::WithZ should be a minimal XS module that just supplements
Math::Clipper, and not a complete copy. The goal is to develop it in that
direction.

Until that's accomplished (and after), the WithZ fork should work the same as 
Math::Clipper, just with the extra Z feature, and initially a different version
of Clipper under the hood.



Changes
Revision history for Perl extension Math::Clipper::WithZ.

1.22  Friday July 05 21:00:00 2013

  - Math::Clipper::WithZ forked from Math::Clipper
  - Early support for Clipper 6.0.0 alpha's Z values in IntPoints
  - Ideally this fork will develop from a full copy of Math-Clipper
    to just a minimal XS module that simply adds the Z feature as
    an run-time option to Math-Clipper.

1.22  Sun May 19 17:14:00 2013

  - Fixed some memory leaks.

1.21  Fri May 10 14:43:00 2013

  - New pt_execute() method exposing PolyTree.
  - New int_offset2()
  - Upgraded the Clipper library to the newest available version 5.1.5.

1.20  Sun Apr 28 15:43:00 2013

  - Fix compilation on Windows.

1.19  Mon Apr 22 13:54:00 2013

  - Upgraded the Clipper library to the newest available version 5.1.4.

1.18  Sun Mar 31 12:46:00 2013

  - new int_offset()
  - new ex_int_offset()
  - new ex_int_offset2()

1.17  Tue Jan 15 14:07:00 2013

  - Expose PFT_POSITIVE and PFT_NEGATIVE constants.
  - Upgraded the Clipper library to the newest available version 5.0.3.

1.16  Wed Dec 26 15:29:00 2012

  - Upgraded the Clipper library to the newest available version 4.10.0.

1.15  Sat Nov 09 17:16:00 2012

  - Upgraded the Clipper library to the newest available version 4.9.6.

1.14  Sat Nov 03 02:00:00 2012

  - Upgraded the Clipper library to the newest available version 4.9.4.

1.13  Wed Nov 01 01:19:00 2012

  - Upgraded the Clipper library to the newest available version 4.9.3.

1.12  Tue Oct 30 01:09:00 2012

  - Fixed Int128 comparison operators in Clipper library, which could 
    cause wrong orientations.

1.11  Mon Oct 22 14:33:00 2012

  - Reduced 32 bit and 64 bit max integer limits introduced in 4.8.6.

1.10  Tue Oct 16 17:39:00 2012

  - Upgraded the Clipper library to the newest available version 4.8.8.

1.08  Mon May 28 19:16:00 2012

  - Upgraded the Clipper library to the newest available version 4.8.3.

1.07  Tue May 01 12:30:00 2012

  - Upgraded the Clipper library to the newest available version 4.8.0.
  - Upgraded the Clipper library to the newest available version 4.7.5.

1.06  Mon Apr 02 18:36:00 2012

  - new simplify_polygon() and simplify_polygons() functions

  - Renamed is_counter_clockwise() to orientation() (the former is still
    exported for backwards compatibility).

1.05  Sun Mar 11 19:06:00 2012

  - Releasing again, since the 1.04 release got out broken probably.

1.04  Sun Mar 11 19:01:00 2012

  - Upgraded the Clipper library to the newest available version 4.7.3.

1.03  Sat Jan 28 16:39:43 2012

  - Added a sub CLONE_SKIP { 1 } to make the module thread-safe in its
    basic form

1.02  Tue Oct 11 16:36:00 2011

  - Upgraded the Clipper library to the newest available version 4.5.5.
    use_full_coordinate_range is always enabled by default, so we now 
    provide this method for legacy purposes but it's actually a no-op.
  
  - Removed warning from integerize_coordinate_sets().

  - Added Test::Deep to prerequisites.

1.01  Mon Oct 03 09:15:00 2011

  - Upgraded the Clipper library to the newest available version 4.4.4.
  
  - New jointype and miterlimit arguments supported for OffsetPolygons

1.00  Mon Sep 19 08:48:00 2011
  * WARNING * This release contains incompatible changes!

  - Upgraded the Clipper library to the newest available version 4.3.0.

  - Clipper 4.3.0 has various interface changes, see the documentation
    for details. Most importantly, the coordinates are now 64bit
    integers instead of floating point numbers.

  - Clipper 4.3.0 now works with large integers internally instead of
    floats. Thus, this release adds support and tests for Clipper's
    int/double conversions.

  - Add ex_execute() for ExPolygons.

  - Add use_full_coordinate_range() method.

  - Remove broken is_clockwise() function and replace with
    is_counter_clockwise(), which is what the C++ function
    IsClockwise() has been doing all along.

  - area() is now a function instead of a method.

  - Exposes polygon_offset function.

  - force_orientation related functions are gone.

  - Compilation fixes for Windows perls and compilers.

  - Internals:
    => Use ExtUtils::Typemaps instead of the deprecated
       ExtUtils::Typemap.
    => Moved all XS++ types to the clipper:: C++ namespace to avoid
       collisions on Win32.

0.01  Thu Aug 19 19:00:00 2010
  
  - original version



src/zfill.h
#ifndef clipper_zfill_h_
#define clipper_zfill_h_

#define ZMARK -1;
#include <iostream>

enum ZFillType { zftNone, zftMax, zftMin, zftMean, zftBothUInt32 };

void zfill_mark(long64 z1, long64 z2, IntPoint& pt) {
  pt.Z = ZMARK;
}

void zfill_mean(long64 z1, long64 z2, IntPoint& pt) {
  pt.Z = (z1 + z2) / 2;
}

void zfill_greater(long64 z1, long64 z2, IntPoint& pt) {
  pt.Z = z1 > z2 ? z1 : z2;
}

void zfill_lesser(long64 z1, long64 z2, IntPoint& pt) {
  pt.Z = z1 < z2 ? z1 : z2;
}

void zfill_first(long64 z1, long64 z2, IntPoint& pt) {
  pt.Z = z1;
}

void zfill_second(long64 z1, long64 z2, IntPoint& pt) {
  pt.Z = z2;
}

// store two 32 bit unsigned ints, stored in one 64 bit int
// if Z is used to hold an index into some other data array
// for each point/edge, this let's us return both indeces
// for the intersection point.
// These indeces would have to be limited to 4,294,967,295 - what fits in a U32 -
// and then any point coming back with a Z > than 4,294,967,295 would obviously
// be an intersection point, and the two indeces could be extracted.
// Should also be able to use this with signed I32 - you would just need
// an extra decoding step to reinterpret the U32 as an I32.

void zfill_both_uint32s(long64 z1, long64 z2, IntPoint& pt) { 

  // Always take the low index if either Z holds a high and a low.
  // The lows should always be the ones relevant to this pt.Z.

  cUInt hi = (cUInt) (z1 & 0xFFFFFFFF);
  cUInt lo = (cUInt) (z2 & 0xFFFFFFFF);
  hi <<= 32;

  // gets cast back to signed integer, but bits should stay the same
  pt.Z = (hi + lo);
}

// The edge order for the z1 and z2 values stored above can be made to
// correspond to edge order in the input and result by swaping the hi and low
// values in these two cases:
// (clip type is NOT intersection or difference) xor (point.Y IS a local extreme)
void zfill_fix_pair_order(IntPoint& prevpt, IntPoint& thispt, IntPoint& nextpt,
  ClipperLib::ClipType ct
  ) {

  // precondition: thispt.Z > 0xFFFFFFFF
  if (true) {
  if ((
       // these four cover union, intersection and difference
       ((thispt.Y > prevpt.Y || (thispt.Y == prevpt.Y && thispt.X > prevpt.X)) 
        && (thispt.Y > nextpt.Y ) // || (thispt.Y == nextpt.Y && thispt.X > nextpt.X))
       )
       ||
       ((thispt.Y > prevpt.Y)
        && (thispt.Y > nextpt.Y || (thispt.Y == nextpt.Y && thispt.X > nextpt.X))
        && ct != ctXor
       )
       || 
       ((thispt.Y < prevpt.Y || (thispt.Y == prevpt.Y && thispt.X < prevpt.X)) 
        && (thispt.Y < nextpt.Y ) // || (thispt.Y == nextpt.Y && thispt.X < nextpt.X))
        )
       || 
       ((thispt.Y < prevpt.Y) 
        && (thispt.Y < nextpt.Y || (thispt.Y == nextpt.Y && thispt.X < nextpt.X)
        && ct != ctXor)
        )
       // XOR (for polygons, not referring to the one below) _might_ be similar 
       // but with X playing the staring role.
       // Or it might be more complicated because it's like an intersection and
       // a difference in one operation. 
       // or
       // Might be able to do it by making the sequence of intersections 
       // consistent, sometimes disambiguating by checking neighbor points -  
       // but that wouldn't happen right in here.
       )
       != // xor
       (ct == ctIntersection || ct == ctDifference)
     ) {    
    thispt.Z = ((thispt.Z & 0xFFFFFFFF) << 32) + (thispt.Z >> 32);    
  }
  }
}

void zfill_both_uint32s_fix_pairs_polygon(ClipperLib::Polygon& poly,
  ClipperLib::ClipType ct) {

  unsigned int len = poly.size();

  if (len > 2) {
    if (poly[len - 1].Z > 0xFFFFFFFF) {
      zfill_fix_pair_order(poly[len - 2], poly[len - 1], poly[0], ct);
    }
    if (poly[0].Z > 0xFFFFFFFF) {
      zfill_fix_pair_order(poly[len - 1], poly[0], poly[1], ct);
    }
    for (unsigned int j = 1; j < len - 1; j++) {
      if (poly[j].Z > 0xFFFFFFFF) {
        zfill_fix_pair_order((poly[j - 1]), (poly[j]), (poly[j + 1]), ct);
      }
    }
  }
}

void zfill_both_uint32s_fix_pairs_polygons(ClipperLib::Polygons& polys,
  ClipperLib::ClipType ct) {

  for (unsigned int i = 0; i < polys.size(); i++) {
    
    zfill_both_uint32s_fix_pairs_polygon(polys[i], ct);

  }

}

void zfill_both_uint32s_fix_pairs_polynode(ClipperLib::PolyNode& polynode,
  ClipperLib::ClipType ct) {

  zfill_both_uint32s_fix_pairs_polygon(polynode.Contour, ct);
  for (int i = 0; i < polynode.ChildCount(); ++i) {
    zfill_both_uint32s_fix_pairs_polygon(polynode.Childs[i]->Contour, ct);
    //Add outer polygons contained by (nested within) holes ...
    for (int j = 0; j < polynode.Childs[i]->ChildCount(); ++j) {
      zfill_both_uint32s_fix_pairs_polynode(*polynode.Childs[i]->Childs[j], ct);
    }
  }

}

void zfill_both_uint32s_fix_pairs_polytree(ClipperLib::PolyTree& polytree,
  ClipperLib::ClipType ct) {

  for (int i = 0; i < polytree.ChildCount(); ++i) {
    zfill_both_uint32s_fix_pairs_polynode(*polytree.Childs[i], ct);
  }

}

// interpret the I64 Z values as 32 bit floats, and store both in one 64 bit int
// May lose precision, but you get two for one.
// Should be useful for low-precision data - 23 bit mantissa, 6 to 9 dec. digits

void zfill_both_float32s(long64 z1, long64 z2, IntPoint& pt) { 

  cUInt hi = (cUInt) (float) z1;
  hi <<= 32;
  cUInt lo = (cUInt) (float) z2;

  // back to signed interger, because that's what Clipper expects
  pt.Z = (long64) lo + hi;

}


void zfill_postprocess(ClipperLib::Polygons& p, ClipType clipType, ZFillType zft) {
    if (zft == zftBothUInt32) {
        zfill_both_uint32s_fix_pairs_polygons(p, clipType);
    }
}

void zfill_postprocess_pt(ClipperLib::PolyTree& p, ClipType clipType, ZFillType zft) {
    if (zft == zftBothUInt32) {
        zfill_both_uint32s_fix_pairs_polytree(p, clipType);
    }
}

void set_zfill_callback(Clipper& THIS, ZFillType zft) {
    switch (zft) {
        case zftNone : THIS.ZFillFunction(0); break;
        case zftMax  : THIS.ZFillFunction(&zfill_greater); break;
        case zftMin  : THIS.ZFillFunction(&zfill_lesser);break;
        case zftMean : THIS.ZFillFunction(&zfill_mean); break;
        case zftBothUInt32 : THIS.ZFillFunction(&zfill_both_uint32s); break;
        default      : THIS.ZFillFunction(0);
    }
}

#endif


src/poly2av.h
#ifndef clipper_poly2av_h_
#define clipper_poly2av_h_

#include "myinit.h"
#include <iostream>

SV*
polygon2perl(pTHX_ const ClipperLib::Polygon& poly)
{
  AV* av = newAV();
  AV* innerav;
  const unsigned int len = poly.size();
  av_extend(av, len-1);
  for (unsigned int i = 0; i < len; i++) {
    innerav = newAV();
    av_store(av, i, newRV_noinc((SV*)innerav));
#ifdef use_xyz
    av_fill(innerav, 2);
#else
    av_fill(innerav, 1);
#endif
    // IVSIZE is from perl/lib/CORE/config.h, defined as sizeof(IV)
#if IVSIZE >= 8
    // if Perl integers are 64 bit, use newSViv()
    av_store(innerav, 0, newSViv(poly[i].X));
    av_store(innerav, 1, newSViv(poly[i].Y));//std::cout << "here with Z [" << poly[i].Z << "]";
#ifdef use_xyz
    av_store(innerav, 2, newSViv(poly[i].Z));
#endif
#else
    // otherwise we expect Clipper integers to fit in the
	// 53 bit mantissa of a Perl double
    av_store(innerav, 0, newSVnv(poly[i].X));
    av_store(innerav, 1, newSVnv(poly[i].Y));
#ifdef use_xyz
    av_store(innerav, 2, newSVnv(poly[i].Z));
#endif
#endif


  }
  return (SV*)newRV_noinc((SV*)av);
}


SV*
polygons2perl(pTHX_ const ClipperLib::Polygons& poly)
{
  AV* av = newAV();
  SV* innerav;
  const unsigned int len = poly.size();
  av_extend(av, len-1);
  for (unsigned int i = 0; i < len; i++) {
    innerav = polygon2perl(aTHX_ poly[i]);
    av_store(av, i, innerav);
  }
  return (SV*)newRV_noinc((SV*)av);
}


SV*
expolygon2perl(pTHX_ const ExPolygon& poly)
{
  HV* hv = newHV();
  (void)hv_stores( hv, "outer", (SV*)polygon2perl(aTHX_ poly.outer) );
  (void)hv_stores( hv, "holes", (SV*)polygons2perl(aTHX_ poly.holes) );
  return (SV*)newRV_noinc((SV*)hv);
}


SV*
expolygons2perl(pTHX_ const ExPolygons& polys)
{
  AV* av = newAV();
  SV* innerav;
  const unsigned int len = polys.size();
  av_extend(av, len-1);
  for (unsigned int i = 0; i < len; i++) {
    innerav = expolygon2perl(aTHX_ polys[i]);
    av_store(av, i, innerav);
  }
  return (SV*)newRV_noinc((SV*)av);
}

SV* polynode2perl(const PolyNode& node);

SV*
polynode_children_2_perl(const PolyNode& node)
{
    AV* av = newAV();
    const unsigned int len = node.ChildCount();
    av_extend(av, len-1);
    for (int i = 0; i < len; ++i) {
        av_store(av, i, polynode2perl(*node.Childs[i]));
    }
    return (SV*)newRV_noinc((SV*)av);
}

SV*
polynode2perl(const PolyNode& node)
{
    HV* hv = newHV();
    if (node.IsHole()) {
        (void)hv_stores( hv, "hole", (SV*)polygon2perl(aTHX_ node.Contour) );
    } else {
        (void)hv_stores( hv, "outer", (SV*)polygon2perl(aTHX_ node.Contour) );
    }
    (void)hv_stores( hv, "children", (SV*)polynode_children_2_perl(node) );
    return (SV*)newRV_noinc((SV*)hv);
}

SV*
polytree2perl(pTHX_ const PolyTree& polytree)
{
    return polynode_children_2_perl(polytree);
}



ClipperLib::Polygon*
perl2polygon(pTHX_ AV* theAv)
{
  const unsigned int len = av_len(theAv)+1;
  ClipperLib::Polygon* retval = new ClipperLib::Polygon(len);
  SV** elem;
  AV* innerav;
  for (unsigned int i = 0; i < len; i++) {
    elem = av_fetch(theAv, i, 0);
    if (!SvROK(*elem)
        || SvTYPE(SvRV(*elem)) != SVt_PVAV
        || av_len((AV*)SvRV(*elem)) < 1)
    {
      delete retval;
      return NULL;
    }
    innerav = (AV*)SvRV(*elem);
    ClipperLib::IntPoint& p = (*retval)[i];
    // IVSIZE is from perl/lib/CORE/config.h, defined as sizeof(IV)
#if IVSIZE >= 8
    // if Perl integers are 64 bit, use SvIV()
    // Clipper.pm then supports 64 bit ints.
    p.X = (ClipperLib::long64)SvIV(*av_fetch(innerav, 0, 0));
    p.Y = (ClipperLib::long64)SvIV(*av_fetch(innerav, 1, 0));
#ifdef use_xyz
    if (av_len(innerav) > 1) {
       p.Z = (ClipperLib::cInt) SvIV(*av_fetch(innerav, 2, 0));
    } // else p.Z defaults to zero
#endif
#else
    // otherwise coerce the Perl scalar to a double, with SvNV()
    // Perl doubles commonly allow 53 bits for the mantissa.
    // So in the common case, Clipper.pm supports 53 bit integers, stored in doubles on the Perl side.
    p.X = (ClipperLib::long64)SvNV(*av_fetch(innerav, 0, 0));
    p.Y = (ClipperLib::long64)SvNV(*av_fetch(innerav, 1, 0));
#ifdef use_xyz
    if (av_len(innerav) > 1) {
       p.Z = (ClipperLib::cInt)SvNV(*av_fetch(innerav, 2, 0));
    } // else p.Z defaults to zero
#endif
#endif
  }
  return retval;
}


ClipperLib::Polygons*
perl2polygons(pTHX_ AV* theAv)
{
  const unsigned int len = av_len(theAv)+1;
  ClipperLib::Polygons* retval = new ClipperLib::Polygons(len);
  SV** elem;
  AV* innerav;
  for (unsigned int i = 0; i < len; i++) {
    elem = av_fetch(theAv, i, 0);
    if (!SvROK(*elem)
        || SvTYPE(SvRV(*elem)) != SVt_PVAV
        || av_len((AV*)SvRV(*elem)) < 1)
    {
      delete retval;
      return NULL;
    }
    innerav = (AV*)SvRV(*elem);
    ClipperLib::Polygon* tmp = perl2polygon(aTHX_ innerav);
    if (tmp == NULL) {
      delete retval;
      return NULL;
    }
    (*retval)[i] = *tmp;
    delete tmp;
  }
  return retval;
}


#define AV_CHECK(outav, hv, key)                                                       \
    STMT_START {                                                                       \
      SV** buf = hv_fetchs(hv, key, 0);                                                \
      if (!buf || !*buf) {                                                             \
        croak("Missing ExPolygon hash key: '" key "' or not an array reference.");     \
      }                                                                                \
      SvGETMAGIC(*buf);                                                                \
      if (SvROK(*buf) && SvTYPE(SvRV(*buf)) == SVt_PVAV) {                             \
        outav = (AV*)SvRV(*buf);                                                       \
      }                                                                                \
      else {                                                                           \
        croak(key " is not an ARRAY reference");                                       \
      }                                                                                \
    } STMT_END

ExPolygon*
perl2expolygon(pTHX_ HV* theHv)
{
  AV* outerav;
  AV* holesav;
  AV_CHECK(outerav, theHv, "outer");
  AV_CHECK(holesav, theHv, "holes");

  ExPolygon* retval = new ExPolygon();
  ClipperLib::Polygon* tmp = perl2polygon(aTHX_ outerav);
  if (!tmp) {
    delete retval;
    return NULL;
  }
  retval->outer = *tmp;

  ClipperLib::Polygons* tmps = perl2polygons(aTHX_ holesav);
  if (!tmps) {
    delete retval;
    return NULL;
  }
  retval->holes = *tmps;

  return retval;
}
#undef AV_CHECK


ExPolygons*
perl2expolygons(pTHX_ AV* theAv)
{
  const unsigned int len = av_len(theAv)+1;
  std::vector<ExPolygon> tmpEx; // Done because of croak

  SV** elem;
  HV* innerhv;
  for (unsigned int i = 0; i < len; i++) {
    elem = av_fetch(theAv, i, 0);
    if (!SvROK(*elem)
        || SvTYPE(SvRV(*elem)) != SVt_PVHV)
      return NULL;
    innerhv = (HV*)SvRV(*elem);
    ExPolygon* tmp = perl2expolygon(aTHX_ innerhv);
    if (tmp == NULL)
      return NULL;
    tmpEx[i] = *tmp;
    delete tmp;
  }

  ExPolygons* retval = new ExPolygons(tmpEx);
  return retval;
}


#endif


lib/Math/Clipper/WithZ.pm
package Math::Clipper::WithZ;

use 5.008;
use strict;
use warnings;
use Carp qw(croak carp);
use Config;

use Exporter();
our $VERSION;
our @ISA = qw(Exporter);

BEGIN {
    use XSLoader;
    $VERSION = '1.22';
    XSLoader::load('Math::Clipper::WithZ', $VERSION);
}

# TODO: keep in sync with docs below and xsp/Clipper.xsp

our %EXPORT_TAGS = (
    cliptypes     => [qw/CT_INTERSECTION CT_UNION CT_DIFFERENCE CT_XOR/],
    #polytypes     => [qw/PT_SUBJECT PT_CLIP/],
    polyfilltypes => [qw/PFT_EVENODD PFT_NONZERO PFT_POSITIVE PFT_NEGATIVE/],
    jointypes     => [qw/JT_MITER JT_ROUND JT_SQUARE/],
    endtypes     => [qw/ET_SQUARE ET_ROUND ET_BUTT ET_CLOSED/],
    zfilltypes     => [qw/ZFT_NONE ZFT_MAX ZFT_MIN ZFT_MEAN ZFT_BOTH_UINT32/],
    utilities       => [qw/area offset is_counter_clockwise orientation integerize_coordinate_sets unscale_coordinate_sets
                    simplify_polygon simplify_polygons int_offset ex_int_offset ex_int_offset2/],
);

$EXPORT_TAGS{all} = [ map { @$_ } values %EXPORT_TAGS ];
our @EXPORT_OK = ( @{ $EXPORT_TAGS{'all'} } );
our @EXPORT = qw();

my %intspecs = (
    '64' => {
            maxint    => 4611686018427387902,   # Clipper-imposed max when using 64 bit integer math
            maxdigits => 19
            },
    '53' => {
            maxint    => 9007199254740992, # signed 53 bit integer max, for integers stored in double precision floats
            maxdigits => 16
            },
    '32' => {
            maxint    => 1073741822,   # Clipper-imposed max to avoid calculations with large integer types
            maxdigits => 10
            },
    );

my $is64safe = ((defined($Config{use64bitint})   && $Config{use64bitint}   eq 'define') || $Config{longsize}   >= 8 ) &&
               ((defined($Config{uselongdouble}) && $Config{uselongdouble} eq 'define') || $Config{doublesize} >= 10);

sub offset {
    my ($polygons, $delta, $scale, $jointype, $miterlimit) = @_;
    $scale      ||= 100;
	$jointype   = JT_MITER if !defined $jointype;
	$miterlimit ||= 2;
	
	my $scalevec=[$scale,$scale];
	my $polyscopy=[(map {[(map {[(map {$_*$scalevec->[0]} @{$_})]} @{$_})]} @{$polygons})];
	my $ret = _offset($polyscopy,$delta*$scale, $jointype, $miterlimit);
	unscale_coordinate_sets($scalevec , $ret) if @$ret;
	return $ret;
	}

*is_counter_clockwise = *orientation;

sub unscale_coordinate_sets { # to undo what integerize_coordinate_sets() does
    my $scale_vector=shift;
    my $coord_sets=shift;
    my $coord_count=scalar(@{$coord_sets->[0]->[0]});
    if (!ref($scale_vector)) {$scale_vector=[(map {$scale_vector} (0..$coord_count-1))];}
    foreach my $set (@{$coord_sets}) {
        foreach my $vector (@{$set}) {
            for (my $ci=0;$ci<$coord_count;$ci++) {
                $vector->[$ci] /= $scale_vector->[$ci] if $scale_vector->[$ci]; # avoid divide by zero
                }
            }
        }
    }

sub integerize_coordinate_sets {
    my %opts=();
    if (ref($_[0]) =~ /HASH/) {%opts=%{(shift)};}
    $opts{constrain} =  1 if !defined($opts{constrain});
    $opts{bits}      = ($is64safe ? 64 : 53) if !defined($opts{bits});
	if ($opts{bits} == 64 && !$is64safe) {$opts{bits} = 53; carp "Integerize to 64 bits requires both long long and long double underlying Perl's default integer and double types. Using 53 bits instead.";}
    $opts{margin} =  0 if !defined($opts{margin});

    # assume all coordinate vectors (points) have same number of coordinates; get that count from first one
    my $coord_count=scalar(@{$_[0]->[0]});

    # return this with scaled data, so user can "unscale" Clipper results
    my @scale_vector;
    
    # deal with each coordinate "column" (eg. x column, y column, ... possibly more)
    for (my $ci=0;$ci<$coord_count;$ci++) {
        my $maxc=$_[0]->[0]->[$ci];
        my $max_exp;

        # go through all the coordinate sets, looking just at the current column
        foreach my $set (@_) {
            # for each "point"
            foreach my $vector (@{$set}) {
                # looking for the maximum magnitude
                if ($maxc<abs($vector->[$ci]) + $opts{margin}) {$maxc=abs($vector->[$ci]) + $opts{margin};}
                # looking for the maximum exponent, when coords are in scientific notation
                if (sprintf("%.20e",$vector->[$ci] + ($vector->[$ci]<0?-1:1)*$opts{margin}) =~ /[eE]([+-])0*(\d+)$/) {
                    my $exp1 = eval($1.$2);
                    if (defined $vector->[$ci] && (!defined($max_exp) || $max_exp<$exp1)) {$max_exp=$exp1} 
                    }
                else {croak "some coordinate didn't look like a number: ",$vector->[$ci]}
                }
            }

        # Set scale for this coordinate column to the largest value that will convert the
        # larges coordinate in the set to near the top of the available integer range.
        # There's never any question of how much precision the user wants -
        # we just always give as much as possible, within the integer limit in effect (53 bit or 64 bit)

        $scale_vector[$ci]=10**(-$max_exp + ($intspecs{$opts{bits}}->{maxdigits} - 1));

        if ($maxc * $scale_vector[$ci] > $intspecs{$opts{bits}}->{maxint}) {
            # Both 53 bit and 64 bit integers
            # have max values near 9*10**(16 or 19).
            # So usually you have 16 or 19 digits to use. 
            # But if your scaled-up max values enter the
            # zone just beyond the integer max, we'll only
            # scale up to 15 or 18 digit integers instead.

            $scale_vector[$ci]=10**(-$max_exp + ($intspecs{$opts{bits}}->{maxdigits} - 2));

            }
        }
    
    # If the "constrain" option is set false,
	# scaling is independent for each
    # coordinate column - all the Xs get one scale
    # all the Ys something else - to take the greatest
    # advantage of the available integer domain.
    # But if the "constrain" option is set true, we use
    # the minimum scale from all the coordinate columns.
    # The minimum scale is the one that will work
    # for all columns, without overflowing our integer limits.
    if ($opts{constrain}) {
        my $min_scale=(sort {$a<=>$b} @scale_vector)[0];
        @scale_vector = map {$min_scale} @scale_vector;
        }

    # Scale the original data
    foreach my $set (@_) {
        foreach my $vector (@{$set}) {
            for (my $ci=0;$ci<$coord_count;$ci++) {
                $vector->[$ci] *= $scale_vector[$ci];
                if (abs($vector->[$ci] < 1)) {$vector->[$ci] = sprintf("%.1f",$vector->[$ci]/10)*10;}
                }
            }
        }

    return \@scale_vector;
    }

# keep this method as a no-op, as it was removed in Clipper 4.5.5
sub use_full_coordinate_range {}

sub CLONE_SKIP { 1 }

1;
__END__

=head1 NAME

Math::Clipper::WithZ - Polygon clipping in 2D

=head1 SYNOPSIS

 use Math::Clipper::WithZ ':all';

 my $clipper_with_z = Math::Clipper::WithZ->new;

 my $p1 = [
    [ 20, -10,    1],
    [ 50,   0,   -2],
    [ 20,  10,   0xFFFFFFFF],
    [-10,   0,    0],
 ];

 $clipper->add_subject_polygon( $p1 );

 my $result = $clipper->execute(CT_UNION);

 # Z value pass through
 # $result is now:
 # [
 #   [
 #     [20, -10, 1],
 #     [50, 0, -2],
 #     [20, 10, 4294967295],
 #     [-10, 0, 0]
 #   ]
 # ]



=head1 DESCRIPTION

C<Clipper> is a C++ (and Delphi) library that implements
polygon clipping. Version 6.0.0 alpha adds compile-time
support for points with an additional Z value, in addition
to the the normal X and Y values.

Z is for arbitrary data associated with each point or edge.
Clipper does nothing with this value but pass it through to 
the result.

For new points generated at intersections, Z is set to 0
by default in results. A "zFill" callback function may be 
provided to set the Z at intersections to something else, 
based on neighboring Z values. But this is a callback 
in C, not Perl.

In addition to providing the basic Z pass-through feature
Math::Clipper::WithZ aims to provide a few predetermined 
callback options that will determine how Clipper sets Z 
values at intersections.

Other than those additional features, this module should 
work identically to Math::Clipper. See the
Math::Clipper docs for overall usage.

=head1 EXPORTS

To specify which predetermined zFill callback to use, this
module defines a few additional constants that may be
imported.

    ZFT_NONE
    ZFT_MAX
    ZFT_MIN
    ZFT_MEAN
    ZFT_BOTH_UINT32

=head1 METHODS

All methods are the same as Math::Clipper except that
the three execute methods take an additional argument,
specifying a predetermined zFill option.

=head2 execute, ex_execute, pt_execute

The optional zFill arguments follows the normal Math::Clipper
arguments.

    my $result = $clipper->execute( CT_UNION , PFT_NONZERO, undef, ZFT_MAX);

This is a union operation, with the subject fill type set to nonzero and the
clip fill type left to it's default.

The last argument specifies that we want to use a zFill function that sets 
Z for the intersection point to the maximum Z value of the two intersecting 
edges. (Points correspond to edges. The Z value of an edge is the Z value of 
the point at the beginning of that edge.)

=head1 Z FILL OPTIONS

=head2 ZFT_BOTH_UINT32

The Z values of the two edges that intersect will be interpreted as
32 bit unsigned integers, and both will be returned.

This could typically be used to identify the input edges that contribute 
to the output intersection point, if the integers are indices into the 
input edge/point list.

Or it could be used to identify which input polygons contribute to an output
polygon, and what subsequences of output polygon points come from which
input polygons. In this case, Z values for all points in each polygon would 
bear the same integer identifying the polygon.

=head2 ZFT_MAX

The intersection point will inherit the maximum Z value of the two 
intersecting edges.

=head2 ZFT_MIN

The intersection point will inherit the minimum Z value of the two 
intersecting edges.

=head2 ZFT_MEAN

The intersection point will inherit the arithmetic mean ((Z1+Z2)/2) of the 
Z values of the two intersecting edges. 

=head2 ZFT_NONE

No callback function will be defined. Clipper will simply set Z to zero for
all intersection points (while still passing through any Z values present in
other points).

=head1 NO Z FILL FOR OFFSETS

Clipper does not attempt to pass Z values through for offset operations.
The way offset points are generated, the result would often be quite scrambled
and meaningless.

However, sometimes it wouldn't. You might consider calculating your own offset points
with Z values in a manner similar to Clipper, and then using a Z-preserving union to
clean up the results, just as Clipper does. It should be fairly simple to glean the 
initial offset point calculation from the Clipper source.

=head1 VERSION

This module was built around, and includes, Clipper version 6.0.0 alpha.

=head1 AUTHOR

Math::Clipper was written by:

Steffen Mueller (E<lt>smueller@cpan.orgE<gt>),
Mike Sheldrake and Alessandro Ranellucci (aar/alexrj)

and these WithZ additions are by Mike Sheldrake.

But the underlying library C<Clipper> was written by
Angus Johnson. Check the SourceForge project page for
contact information.

=head1 COPYRIGHT AND LICENSE

The C<Math::Clipper::WithZ> module is

Copyright (C) 2013 Mike Sheldrake

And is derived, with only a few additions, from C<Math::Clipper>

Copyright (C) 2010, 2011 by Steffen Mueller

Copyright (C) 2011 by Mike Sheldrake

Copyright (C) 2012, 2013 by Alessandro Ranellucci and Mike Sheldrake

but we are shipping a copy of the C<Clipper> C++ library, which
is

Copyright (C) 2010, 2011, 2012, 2013 by Angus Johnson.

C<Math::Clipper::WithZ> is available under the same
license as C<Clipper> itself. This is the C<boost> license:

  Boost Software License - Version 1.0 - August 17th, 2003
  http://www.boost.org/LICENSE_1_0.txt
  
  Permission is hereby granted, free of charge, to any person or organization
  obtaining a copy of the software and accompanying documentation covered by
  this license (the "Software") to use, reproduce, display, distribute,
  execute, and transmit the Software, and to prepare derivative works of the
  Software, and to permit third-parties to whom the Software is furnished to
  do so, all subject to the following:
  
  The copyright notices in the Software and this entire statement, including
  the above license grant, this restriction and the following disclaimer,
  must be included in all copies of the Software, in whole or in part, and
  all derivative works of the Software, unless such copies or derivative
  works are solely in the form of machine-executable object code generated by
  a source language processor.
  
  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE, TITLE AND NON-INFRINGEMENT. IN NO EVENT
  SHALL THE COPYRIGHT HOLDERS OR ANYONE DISTRIBUTING THE SOFTWARE BE LIABLE
  FOR ANY DAMAGES OR OTHER LIABILITY, WHETHER IN CONTRACT, TORT OR OTHERWISE,
  ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
  DEALINGS IN THE SOFTWARE.

=cut
